### ISSUE TITLE
New Version

### ISSUE THREAD

**Issue by gbrosenthal** [JorgenPhi/php-snapchat#89]
In snapchat_agent the version needs to be changed to '9.0.1'

Additionally, has anyone been able to post stories since the update?

**Comment by Snowlav** [comment]
I'm on the 9.0.2.1 beta and a lot has changed, though posting stories shouldn't give any issues.
how are you uploading stories?

**Comment by gbrosenthal** [comment]
$snapchat = new Snapchat('uname', 'password');
$id = $snapchat->upload(
     Snapchat::MEDIA_IMAGE,
     file_get_contents('/path/to/image.png')
);
$snapchat->send($id, array('uname'));

**Comment by liamcottle** [comment]
If you are using the 9.0+ user agent, you need to make sure you have
implemented the new Login parameters.

Check out my working 9.0.2.0 login request here:

http://github.com/liamcottle/snaplib

**Comment by gbrosenthal** [comment]
Logging in isnt the issue. Im able to get users etc... so i know im authenticated

**Comment by cuonic** [comment]
Make sure you send all the following data correctly. This hasn't changed at all compared to 8.1.1

# /bq/retry_post_story

#### Description
Used to post a story

#### Request
* `my_story` - (Boolean)
* `time` - Time to display the snap (Float)
* `group_ids`- (Array)
* `media_id` - `[USERNAME]~[RANDOM UUID]` (String)
* `story_timestamp` - (Timestamp)
* `zipped` - `0` or `1` (Int)
* `caption_text_display` - (String)
* `client_id` - Same as `media_id` (String)
* `type` - Media type (Int)
* `data` - Encoded snap data
* `req_token`
* `timestamp`
* `username`

#### Response
* `json` - (Array)
    * `story` - (Array)
        * `caption_text_display` - (String)
        * `client_id` : Same as `media_id` (String)
        * `id` - `[USERNAME]~[TIMESTAMP]` (String)
        * `mature_content` - (Boolean)
        * `media_id` - `[USERNAME]~[RANDOM UUID]` (String)
        * `media_iv` - Base64 encoded encryption IV for the snap data (String)
        * `media_key` - Base64 encoded decryption key for the snap data (String)
        * `media_type` - Media type (Int)
        * `media_url` - A URL to fetch the story data, generally `https://feelinsonice-hrd.appspot.com/bq/story_blob?story_id=[media_id]` (String)
        * `thumbnail_iv` - Base64 encoded encryption IV for the thumbnail image (String)
        * `thumbnail_url` - A URL to fetch the thumbnail image, generally `https://feelinsonice-hrd.appspot.com/bq/story_thumbnail?story_id=[media_id]` (String)
        * `time` - Number of seconds to display snap (Float)
        * `time_left` - Time left before expiration of the story (Timestamp)
        * `timestamp` - Time posted (Timestamp)
        * `username` - Username of the user who posted the story (String)
        * `zipped` - (Boolean)

**Comment by gbrosenthal** [comment]
Sorry, just realized that my example was sending to single users. But i havent been able to do that either haha.

**Comment by gbrosenthal** [comment]
But it also looks like a lot of those variables are not set by this library. Ie not in 'function setStory' in the snapchat class, nor 'function post' in the snapchat_agent one

**Comment by cuonic** [comment]
The library is not exactly up to date, if you want it to work you need to know what the actual application sends to the API, and then you have to add / remove parameters to each request in the library.

Here is some API documentation : https://github.com/JorgenPhi/php-snapchat/wiki/API-v2-Research

**Comment by Snowlav** [comment]
I'm getting real tired of snapchat and their api updates, I'll tell you that much

**Comment by cuonic** [comment]
Well the API updates are a good thing, it brings new features to the application for us to use and abuse. We just need to keep up with constantly analyzing the application's API calls.

**Comment by Snowlav** [comment]
The latter is hard though, I do not mind it actually, but snapchat is not only adding user features but I have the feeling they are also making it harder for us to abuse every update.

**Comment by Scape373** [comment]
Any fix for this yet? 
I can login and update snaps and save snaps, but cannot upload or set a story 
This literally worked yesterday with no change. I tried updating my user agent to 9.0.2, and 9.0.1 and still doesnt work.. What am i missing?

$final1 = "http://-snip-watermark.php?filename=$value";
$id = $snapchat->upload(
Snapchat::MEDIA_IMAGE,
file_get_contents($final1)); 
$snapchat->setStory($id,5);

**Comment by gbrosenthal** [comment]
I've been able to upload (get a 200 Response), but I havent been able to send single snaps or set my story. If anyone has been able to get this library working in the last couple days, please share.

**Comment by Snowlav** [comment]
Are you uploading a JPG image ?
What responses do you get?

Can you elaborate "I can login and update snaps and save snaps, but cannot upload or set a story. It just uploads a blank gray thumbnail and disappears when i try to click on it"
Do you mean on the application?

**Comment by Snowlav** [comment]
Remember this lib is outdated guys, if you are mixing new headers with old endpoints, or old endpoints with new endpoints etc, it will not work.
Which endpoint are you guys using to login and which are you using to set story.

In the new versions of snapchat you no longer need to use upload and send but simply post to story.

We need more information than that is given right now.

**Comment by JorgenPhi** [comment]
Speaking of outdated, if anyone wants access to accept pull changes and
write to the repo directly, let me know. I don't have the time to keep this
relevant.

**Comment by Scape373** [comment]
So what you are saying is my code should look more like this now?:

It's still not working.. :(

Gets contents of folder 
uploads the images in the folder

$directory = 'saved';
$scanned = array_diff(scandir($directory), array('..', '.'));
   foreach($scanned as $value){
      //Watermark
      $snapchat->setStory($value,5);
      var_dump($value);
   }

$value comes out as a string:
string '6.jpg' (length=5)

**Comment by mgp25** [comment]
Hey guys i have updated the login function and adding the new features https://github.com/mgp25/Snap-API.

Most of the features and improvements comming soon.

Regards

**Comment by cuonic** [comment]
Logging in with 9.0.2.0's API is pratically impossible for the Android version. They have started checking device tokens, so you need to fetch a valid device token before logging in, and there is a new "Authorization" header containing a Google OAuth2 token....

**Comment by Snowlav** [comment]
@cuonic isn't the google oath2 token optional, and by device token do you mean a device ID ?
These can be generated pretty easy right ?

**Comment by cuonic** [comment]
It's no longer option, if you don't provide the "ptoken" on login you 
get "401 UNAUTHORIZED", if you create one randomly you get "Oh no! Your 
login temporarily failed, so please try again later. If your login 
continues to fail, please visit 
https://support.snapchat.com/a/failed-login :)"

The message doesn't change when adding or removing the Authorization header

**Comment by Snowlav** [comment]
hmm so we would need to fetch valid info first using fiddlr or something alike, and once logged in store auth token, that is if we are using the 9.0.2.0 api.
Is this correct, or not possible?

**Comment by cuonic** [comment]
Yeah, but that's not really a solution, you shouldn't have to run the 
official application along side just to get tokens, we need to emulate this.

**Comment by Snowlav** [comment]
I like this challenge

**Comment by Snowlav** [comment]
But I question if it's feasible

**Comment by Snowlav** [comment]
is it possible to bring out an api for a previous no longer supported by snapchat version, without getting our accounts banned?

**Comment by cuonic** [comment]
Well using 8.1.1's API for the moment will probably work for a while, 
but then they will kill it.

**Comment by Snowlav** [comment]
Any other options? Long term solutions?

**Comment by hako** [comment]
I've seen the OAuth thing before it became mandatory in the beta, AFAIK it looks like signing up does not have to use OAuth though. And the `ptoken` is new to me.

@Snowlav We can only look inside the APK I guess, unless there are other solutions.

**Comment by Snowlav** [comment]
@hako inside the APK you mean modding the apk or reverse engineering it in the hopes of finding a solution?

**Comment by liamcottle** [comment]
`ptoken` is the Google Cloud Messaging Registration ID, it's the same value that you post to `/ph/device` in the `device_token` param. It's used by the snapchat server to send notifications to your device via GCM.

As for logging in, I have noticed they have made the header `Authorization` mandatory. Also, it must be a valid one...

If you pass a blank value in the `Authorization` header, you'll get a 401, the snapchat server checks the Google servers (I guess) to check if the Bearer value is legit, if not 401...

I have made a successful login by using an already created bearer for the `Authorization` header from a snapchat app login request.

So, if we can make valid value for the `Authorization` header, we can get in.

**Comment by cuonic** [comment]
I'm working on generating the device token signature  for login at the 
moment, it's pretty simple. The Authorization header is an OAuth2 token 
from Google based on Snapchat's client secret

**Comment by liamcottle** [comment]
Ok,

So did a bit of digging, and because the OAuth token is tied to the users Google Account, what I did was remove my local Google Accounts, and then perform a snapchat login.

I noticed that in the `Authorization` header that if you don't have a Google Account connected, the official Snapchat app sends `Bearer ng` as the `Authorization` header...

Doing this still returns a 401 if you pass in a blank `ptoken` which is a GCM reg id...

Sending a blank or random `ptoken` also fails...

So, what snapchat server is doing, is if it gets `Bearer ng` as the `Authorization` header it asks Google if the notification ID passed in `ptoken` is correct...

I was able to make a successful login request by sending:
Header: `Authorization: Bearer ng`
Param: `ptoken` with an actual Google GCM ID.

**Comment by Snowlav** [comment]
can Google GCM IDs be generated in bulk ?

Thanks for all the research to both you and cuonic btw :)

**Comment by cuonic** [comment]
Ideally we need to be able to generate both the GCM and Google OAuth 
token outside of Android, I've got all the POST parameters for these 
requests, I'll take a look

**Comment by liamcottle** [comment]
A Google GCM ID is specific to the device and application that it is requested for, the device itself has to request it from the Google Server.

I've just factory reset one of my S5 test devices so I can check GCM registration requests without any old data being sent.

Might be able to create some "virtual" device IDs, that work...

See what I can do...

**Comment by ddaddy** [comment]
Just listening from the outskirts.
Have you considered how it must be different for iOS? They don't have GCM so completely different id's and from what I know, the apple messaging id's can't be confirmed.

**Comment by liamcottle** [comment]
I thought about this a little while ago, I'll just setup an iOS device with proxy stuff, and see what I get.

**Comment by liamcottle** [comment]
Using the iPad user Agent

Snapchat on iOS sends `X-Snapchat-Client-Token` which is different on each login request.

Headers:
![screen shot 2015-02-01 at 12 07 19 pm](https://cloud.githubusercontent.com/assets/3173581/5990212/0b5c3c34-aa0b-11e4-8b53-d848cd89fc43.png)

Params:
![screen shot 2015-02-01 at 12 07 48 pm](https://cloud.githubusercontent.com/assets/3173581/5990213/0b6227e8-aa0b-11e4-8a49-8d27cb8394a8.png)

... because I don't know how to reverse engineer iOS apps, I can't see how the `X-Snapchat-Client-Token` is generated... Anyone know how to do so? Or is it not possible...

**Comment by mgp25** [comment]
@liamcottle its possible but its harder to reverse it, its better to reverse android apk, im uploading tomorrow a patched apk for you guys, it will print everything in logcat and it will help us in next releases :)

Regards

**Comment by Snowlav** [comment]
Together we can do this :D

**Comment by liamcottle** [comment]
100%

**Comment by liamcottle** [comment]
Thanks to who ever changed the passwords to my test accounts... You should've made your own... Just slows down what I'm doing...

**Comment by cuonic** [comment]
@liamcottle

`dtoken1i` is the Device token ID
`dtoken1v` is the Device token value
`dsig` is the Device token signature

`dsig` is the SHA256 HMAC of `[USERNAME]|[PASSWORD]|[TIMESTAMP]|[REQUEST_TOKEN]` using `dtoken1v` as the secret key.

You can obtain a new Device token set (`dtoken1i` and `dtoken1v`) by posting to `/loq/device_id` without authentication, only a timestamp is required

    public function getDeviceToken() 
    {
        $timestamp = parent::timestamp();

        $result = parent::post(
            '/loq/device_id',
            array(
                'timestamp' => $timestamp,
            ),
            array(
                parent::STATIC_TOKEN,
                $timestamp,
            ));

        return $result;
    }

The signature is generated by the following function in the 9.0.2.0 Android application : 

    public String a(DeviceToken devicetoken, String s, String s1, String s2, String s3)
    {
        SecretKeySpec secretkeyspec = new SecretKeySpec(devicetoken.b().getBytes(Charsets.UTF_8), "HmacSHA256");
        String s4;
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(secretkeyspec);
        s4 = (new String(Hex.encodeHex(Arrays.copyOf(mac.doFinal((newStringBuilder()).append(s).append("|").append(s1).append("|").append(s2).append("|").append(s3).toString().getBytes(Charsets.UTF_8)), 10)))).toLowerCase(Locale.ENGLISH);
        return s4;
        Object obj;
        obj;
        _L2:
        Timber.a("DeviceTokenUtils", ((Throwable) (obj)));
        return null;
        obj;
        if (true) goto _L2; else goto _L1
        _L1:
    }

`s` is the username
`s1` is the password
`s2` is the timestamp
`s3` is the request token

`devicetoken.a()` is the Device token ID
`devicetoken.b()` is the Device token value

Also I've noticed that the timestamp format has changed with the newer versions, there are now 13 digits (Example: `1422744159226`)

**Comment by liamcottle** [comment]
I just finished coding this in Java ;)

I believe they are using `System.currentTimeMillis();` instead of `new Date().getTime();`

**Comment by cuonic** [comment]
@liamcottle I'm trying to implement the hashing function in PHP but I'm not able to reproduce all the getBytes, and I have no idea how the mac function works on byte arrays.

**Comment by cuonic** [comment]
@liamcottle Could you send me an example of all the different variables (username, password, timestamp, request token), the `dtoken1v` and the `dsig` for the specified timestamp and request token so that I can keep trying the HMAC function until it produces the correct output ?

**Comment by liamcottle** [comment]
From Login Request:   
    dtoken1i: 1bd3c0fb-8c5a-4633-acfa-8f96cda1a405
    dtoken1v: neu8H3+7TgDwq7GUcmwSpQ==

Next Login with DSIG:

    dsig: bd05d90bef579a1b744d
    dtoken1i: 1bd3c0fb-8c5a-4633-acfa-8f96cda1a405
    req_token: 93099f5b6a1136b86ea04ee5fcad85d4b9e64d8b190cda8119b48614d0c510db
    timestamp: 1422755390239
    username: govjnwogl

**Comment by cuonic** [comment]
Could you provide the password as well, it's used to create the `dsig`. 
You can send it via email if you want (cuonic@cuonic.com). Promise I 
will not use it or change it :)

**Comment by liamcottle** [comment]
Yeah sorry, forgot to copy it:

    Username: govjnwogl
    Password: snapchat-test-1

**Comment by cuonic** [comment]
@liamcottle Thank you very much :)

**Comment by liamcottle** [comment]
This works mint:

    <?php

    $dtoken1i = "1bd3c0fb-8c5a-4633-acfa-8f96cda1a405";
    $dtoken1v = "neu8H3+7TgDwq7GUcmwSpQ==";
    $req_token = "93099f5b6a1136b86ea04ee5fcad85d4b9e64d8b190cda8119b48614d0c510db";
    $timestamp = "1422755390239";
    $username = "govjnwogl";
    $password = "snapchat-test-1";

    $expected_dsig = "bd05d90bef579a1b744d";

    $dsig = hash_hmac('sha256', $username . "|" . $password . "|" . $timestamp . "|" . $req_token, $dtoken1v);
    $dsig = substr($dsig, 0, 20);

    echo "Expected: " . $expected_dsig;
    echo "<br/>";
    echo "Result: " . $dsig;

    ?>

**Comment by cuonic** [comment]
@liamcottle Yeah I just worked that out, I thought my HMAC was wrong but the API is still rejecting me, now I need to implement the ptoken and Authorization header.

**Comment by liamcottle** [comment]
Chuck this in the `ptoken` param:
    APA91bE61mGyo97nprk1Gq_-OvKs6FxFXhuIGmoILx2-Tr960QVeKWrBKsWikG5znzkGJOXF9Udz4pd-TGJnSbfXEdPyNN_CLhiWkBtNyA0adZ8NPXA9B9PXatVTHmxfnjcaIBumpsyUdRNgYYJcc4xoq5M_k65nKw

... and then `Bearer ng` as the `Authorization` header and see if that works...

Only use that as the `ptoken` to see if it works... As it is a valid one.

**Comment by cuonic** [comment]
@liamcottle It works perfectly, thanks. Now it's time to generate these.

**Comment by liamcottle** [comment]
Mint!

One sad thing, is that `ptoken` requires an Android device to make a request to Google with the GooglePlayServices library...

I'm not sure how we can go about doing this...

**Comment by cuonic** [comment]
Yeah I'm not sure that we can get these as easily. i'll try keeping to `Bearer ng` for the moment with a valid GCM token. I've tried randomly generating these but they are also verified.

**Comment by liamcottle** [comment]
Just made a GCM Registration request, via an android test app with a random package name eg com.example.randomapp and gave the Registration ID to `ptoken` snapchat gives a 401.

... but if I clone the request that was made to register for GCM, and change the package name in the request, to com.snapchat.android then give that Registration ID to snapchat, then it logs in successfully...

So, they're making sure that it's the legit snapchat app that is requesting access to notification system...

Rather annoying, that'll be why notifications no longer work in my SaveMySnaps app...

--

Also, we only need to provide one form of Google Authentication while logging in...

If we give `Authorization: Bearer ng` then we **must** provide a valid GCM Registration ID (via `ptoken`)... Otherwise we get a 401...

However, if we provide a valid Google Authorization Bearer, (not value as "ng") then, we don't have to pass in `ptoken` at all while logging in... (This is the behaviour in the official snapchat app...)

--

So I'm trying to find a Google API that can be called which gives us a valid Authorization value that can be used...

**Comment by cuonic** [comment]
@liamcottle Emulating the Google OAuth token request is simple, the request is always the same for a given user

## Headers

    device:           3730d21d8976a879
    app:              com.snapchat.android
    User-Agent:       GoogleAuth/1.4 (mako JDQ39)
    content-length:   548
    content-type:     application/x-www-form-urlencoded
    Host:             android.clients.google.com
    Connection:       Keep-Alive
    Accept-Encoding:  gzip

## Body

    device_country:                fr
    operatorCountry:               fr
    lang:                          fr_FR
    sdk_version:                   17
    google_play_services_version:  6599036
    accountType:                   HOSTED_OR_GOOGLE
    Email:                         google.play.associated.email@gmail.com
    service:                       audience:server:client_id:694893979329l59f3phl42et9clpoo296d8raqoljl6p.apps.googleusercontent.com
    source:                        android
    androidId:                     3730d21d8976a879
    app:                           com.snapchat.android
    client_sig:                    49f6bgdb81d89a9e38d65de76f09355071bd67e7
    callerPkg:                     com.snapchat.android
    callerSig:                     49f6bgdb81d89a9e38d65de76f09355071bd67e7
    EncryptedPasswd:               oauth2rt_1/kJBxEteTjVjdB41EIWXZPtEHmWqHtMtPgm55rZNvXQA

All of this can be intercepted once using mitmproxy, and once hard coded into the application, it's all good. You could essentially create a dummy account just for generating OAuth tokens.

Proof of concept code : 

    private function getAuthToken() {

		$ch = curl_init();

		$postfields = array(
			'device_country' => 'fr',
			'operatorCountry' => 'fr',
			'lang' => 'fr_FR',
			'sdk_version' => '17',
			'google_play_services_version' => '6599036',
			'accountType' => 'HOSTED_OR_GOOGLE',
			'Email' => 'google.play.associated.email@gmail.com',
			'service' => 'audience:server:client_id:694893979329-l59f3phl42et9clpoo296d8raqoljl6p.apps.googleusercontent.com',
			'source' => 'android',
			'androidId' => '3730d21d8976a879',
			'app' => 'com.snapchat.android',
			'client_sig' => '49f6bgdb81d89a9e38d65de76f09355071bd67e7',
			'callerPkg' => 'com.snapchat.android',
			'callerSig' => '49f6bgdb81d89a9e38d65de76f09355071bd67e7',
			'EncryptedPasswd' => 'oauth2rt_1/kJBxEteTjVjdB41EIWXZPtEHmWqHtMtPgm55rZNvXQA'
		);

		$headers = array(
			'device: 3730d21d8976a879',
			'app: com.snapchat.android',
			'User-Agent: GoogleAuth/1.4 (mako JDQ39)',
			'Accept-Encoding: gzip'
		);

		curl_setopt($ch, CURLOPT_URL, "https://android.clients.google.com/auth");
		curl_setopt($ch, CURLOPT_POST, TRUE);
		curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($postfields));
		curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
		curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, FALSE);
		curl_setopt($ch, CURLOPT_ENCODING, "gzip");
		curl_setopt($ch, CURLINFO_HEADER_OUT, true);

		$result = curl_exec($ch);

		if(curl_getinfo($ch, CURLINFO_HTTP_CODE) != 200) {
			$return['error'] = 1;
			$return['data'] = $result;

			return $return;
		}

		curl_close($ch);

		$return['error'] = 0;
		$return['auth'] = substr(explode("\n", $result)[1], 5);

		return $return;
	}

**Comment by liamcottle** [comment]
That all looks good man, I'll setup some dummy accounts, and intercept the request so I can grab the encrypted password.

Then setup up a test token server, so I can pull an auth token remotely.

**Comment by ddaddy** [comment]
I know your working with the Android app, but on iOS if you say NO to accepting notifications, the returned push ID is 0, so they must have a way to process these also. Is there a way on Android to turn off the notifications and therefore not get a GCM? Or is one always returned?

**Comment by LinoSuarez** [comment]
On Snapchat/9.0.2.0 you're able to log in with only:

* username
* password
* req_token
* timestamp

The problem is this thing: "X-Snapchat-Client-Token", we need to find out how is generated. Right now I'm reversing the app I will post any news. Good luck :+1:

**Comment by cuonic** [comment]
GCM Requests can also be replayed using similar code, you just need to sniff one real request then it's easy : 

    public function getGCMToken() {

		$ch = curl_init();

		$postfields = array(
			'X-GOOG.USER_AID' => '3976909495270287464',
			'app' => 'com.snapchat.android',
			'sender' => '191410808404',
			'cert' => '49f6badb81d89a9e38d65de76f09355071bd67e6',
			'device' => '3976909495270287464',
			'app_ver' => '507',
			'info' => '',
		);

		$headers = array(
			'app: com.snapchat.android',
			'User-Agent: Android-GCM/1.4 (mako JDQ39)',
			'Authorization: AidLogin 3976909495270287464:285036945524349568'
		);

		curl_setopt($ch, CURLOPT_URL, "https://android.clients.google.com/c2dm/register3");
		curl_setopt($ch, CURLOPT_POST, TRUE);
		curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($postfields));
		curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
		curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, FALSE);
		curl_setopt($ch, CURLOPT_ENCODING, "gzip");
		curl_setopt($ch, CURLINFO_HEADER_OUT, true);

		$result = curl_exec($ch);

		if(curl_getinfo($ch, CURLINFO_HTTP_CODE) != 200) {
			$return['error'] = 1;
			$return['data'] = $result;

			return $return;
		}

		curl_close($ch);

		$return['error'] = 0;
		$return['token'] = substr($result, 6);

		return $return;
	}

Using the two functions I provided you can produce a valid Google OAuth2 token and a valid Google Cloud Messaging token, making your login work every time :)

**Comment by LinoSuarez** [comment]
Here is the latest Snapchat source code.
[Download](https://mega.co.nz/#!BkImGbIJ!jccGJmy4R6bbR4pPMrdk0Rjb_-7xiZXZNv6HiuufGKo)

It includes latest Snapchat.sdk and source code in Java, enjoy :facepunch:

**Comment by cuonic** [comment]
@LinoSuarez Where are you seeing "X-Snapchat-Client-Token" being sent as a header ?

**Comment by LinoSuarez** [comment]
@cuonic it's on the iPhone version of the app. The login it's much more simple than on android the unique problem is that Client token... I'm working on reversing the app but it's much more difficult than on android, its taking me some time

**Comment by cuonic** [comment]
I've put all my code here : https://github.com/cuonic/php-snapchat

This includes logging in, getting updates, sending a device token, getting a GCM token, getting a Google OAuth2 token and downloading a Snaptag image

**Comment by mgp25** [comment]
I recently added a useful debug mode, if anyone wants to collaborate, feel free to send pull request. Within the next few days, ill be posting in the wikis repo all the information about:

- reversing
- using the api
- full disclosure
- patched apks
- iphone information about snapchat.

I'll add PSR-2 coding style once i write all documentation.

https://github.com/mgp25/Snap-API

P.S: Snapchat, we love playing the cat-and-mouse game :)

Regards

**Comment by LinoSuarez** [comment]
@cuonic  I get 
array (size=1)
  'message' => string 'Failed to get new Device token set.' (length=35)
Any idea? :/

**Comment by liamcottle** [comment]
Did you replace all of the place holder values?

**Comment by LinoSuarez** [comment]
@liamcottle Ah I just saw that, thought it was ready to go hahah. Any way to get all those values without sniffing snapchat?

**Comment by cuonic** [comment]
Sniffing the real values is required at least once, otherwise we would 
have to reverse engineer Google's OAuth token generating request, and 
I'm sure all have better things todo than throw ourselves into that.

**Comment by LinoSuarez** [comment]
@cuonic What about asking the user to enter google acc then use Google's OAuth API to get those values?

**Comment by cuonic** [comment]
Because this is Google's OAuth API made specifically for authenticating 
android devices, it's not just basic OAuth, so it requires a bit more 
than a username and password

**Comment by LinoSuarez** [comment]
@cuonic Got it, so how it would be possible now to scale this API, I mean could we just use the same google acc for everyone using the app? Would that be to suspicious?

**Comment by cuonic** [comment]
It would probably be suspicious, and Snapchat will be able to find out, 
every OAuth2 token gives them basic info about your Google Play account 
(email and other stuff) so if they see 10+ user's using the same Google 
Play account it'll start to look fishy.

**Comment by LinoSuarez** [comment]
@cuonic I know... Maybe the solution would be to add a proxy to the API witch can read that information needed from Google Play. Otherwise I don't see how this could scale up. Or maybe try to implement iOS login.

**Comment by cuonic** [comment]
You could create a quick android app to fetch an auth token pretending 
to be Snapchat, if you use all their API keys it should be good.

**Comment by mgp25** [comment]
Im working with iOS login, ill have it ready in the next few days, also working in a better way to extract it from android device.

**Comment by LinoSuarez** [comment]
@mgp25 Have you find out how X-Snapchat-Client-Token is generated?

**Comment by liamcottle** [comment]
I reckon figuring out how that X-Snapchat-Client-Token is generated... because if we get that, we don't need to do any of the notification tokens, or google oauth bearer.... because we can say no to ios notifications, and it won't send that param...

**Comment by mgp25** [comment]
@LinoSuarez i haven't checked that, but it should be easy to find. I'll work on that today.

**Comment by cuonic** [comment]
Definetely more interesting to use iOS for logging in, the rest of the 
requests should be the same afterwards right ?

**Comment by mgp25** [comment]
@cuonic Yes, are the same, only login values were different.

**Comment by liamcottle** [comment]
Yep, just sniffed some iOS requests...

They are the same as android other than the login request, and that the `User Agent` is the iOS one... and they also pass `X-Timestamp` header which is the same as the timestamp passed as a post param

**Comment by mgp25** [comment]
I have it almost ready. If there is someone experienced in reversing (smali, android, etc..) and have android device, please contact with me.

Regards

**Comment by KyleBoyer** [comment]
If I have no android device, should I hold off on updating my useragent from 8.1.1 to the 9.0.2.0 until the iOS version is figured out? Otherwise, is there a way I can get all that data without an android device on me?

**Comment by Snowlav** [comment]
You could try genymotion or bluestacks with proxycap.
Genymotion I am pretty sure about but bluestacks not 100%

**Comment by Snowlav** [comment]
@cuonic I am not sure if everything is the same after logging in, when I find my iphone charger I'll so some comparisons

**Comment by Snowlav** [comment]
Nvm I missed the replies you got, my bad ?

**Comment by KyleBoyer** [comment]
Using VirtualBox, am I able to generate the following:
			'X-GOOG.USER_AID' => 'YOUR_GOOGLE_USER_AID',
			'sender' => 'YOUR_SENDER_KEY',
			'cert' => 'YOUR_CERT_KEY',
			'device' => 'YOUR_DEVICE_ID',
			'Authorization: AidLogin YOUR_GOOGLE_USER_AID'
And:
			'accountType' => 'HOSTED_OR_GOOGLE',
			'Email' => 'GOOGLE_PLAY_ASSOCIATED_EMAIL',
			'androidId' => 'YOUR_ANDROID_ID',
			'client_sig' => 'YOUR_CLIENT_SIG',
			'callerSig' => 'YOUR_CALLER_SIG',
			'EncryptedPasswd' => 'YOUR_ENCRYPTED_PASSWORD'

???

**Comment by mgp25** [comment]
@KyleBoyer im researching for a better solution for android using snapchat token :) but if you have android device, you can capture that request, i'll a guide tonight in my repo.

Regards

**Comment by LinoSuarez** [comment]
@mpg25 You should rally checkout iOS

**Comment by LinoSuarez** [comment]
*Really

It looks so much easier, but now I have exams so don't have that much time to reverse it

**Comment by mgp25** [comment]
@LinoSuarez Yes, i already doing that, i already decrypted the ipa and very easy to follow it will take a little but yeah i can figure it out :smile:

**Comment by LinoSuarez** [comment]
Could you send me the decrypted ipa? Idk why my iPhone now crashes now while I'm deofuscating it 😪

**Comment by KyleBoyer** [comment]
Can I also get a copy @mgp25, my iDevice is not jailbroken :X

**Comment by liamcottle** [comment]
Here's the decrypted binary, (note that it's the Snapchat binary inside Snapchat.app and not all the other stuff).

Snapchat Binary:
`http://liamcottle.com/stuff/snapchat.tar.gz`

Snapchat Classes:
`http://liamcottle.com/stuff/snapchat-class.txt`

You can use `gdb` to insert breakpoints in code etc.

I'm trying to find how they're generating the X-Snapchat-Client-Token

Hopefully you guys might be able to find it first...

**Comment by cuonic** [comment]
Well this just happened trying to use the new API : 
"Oh no! Your account has been locked for repeatedly using third-party applications. For more information, or if you feel we've made a mistake, please visit https://www.snapchat.com/locked"

I'm trying everything to not get detected, but somehow they still manage to detect 3rd party usage

**Comment by liamcottle** [comment]
Hmm...

Could be that your server timestamp is different?
Are you sending the Keeping the connection alive? `Keep-Alive`

I'm still working on the decrypted iOS code....

**Comment by cuonic** [comment]
Server timestamp is good, good point about the Keep-Alive, I'll try that.

**Comment by LinoSuarez** [comment]
Do you guys know any way to sniff my MAMP server? I mean it's not that hard to jut add some debug code, but it's always easier to see the packages on Charles

**Comment by liamcottle** [comment]
What are you needing to do that for? Also, you can add enable Mac OSX proxy via Charles.

**Comment by LinoSuarez** [comment]
It doesn't seem to work charles and MAMP don't play nicely together, it's easier to see your request on Charles and just compare them to the original request and if something it's not working correctly then you see what are the differences on the requests.

**Comment by liamcottle** [comment]
Ahh makes sense, forgot this was the PHP lib... sorry.

**Comment by LinoSuarez** [comment]
@JorgenPhi Ah true, thanks!

**Comment by JorgenPhi** [comment]
Don't forget CURLOPT_SSL_VERIFYHOST & CURLOPT_SSL_VERIFYPEER need to be 0.

**Comment by KyleBoyer** [comment]
On iOS when snapchat opens/closes it sends requests to: https://sc-analytics.appspot.com
The content of these requests include:
I'm not sure what content could be maliciously, so I censored out a lot of variables. 
https://sc-analytics.appspot.com/analytic_events
[{
	"session_id": #############,
	"event_name": "session_start",
	"client_upload_ts": #############,
	"event_id": #####,
	"client_ts": #############
}]
and:
https://sc-analytics.appspot.com/post_events
{
	"common_params": {
		"user_id": "CENSORED",
		"city": "Unimplemented",
		"sc_user_agent": "Snapchat\/9.1.0.10 (iPhone6,1; iOS 8.1.3; gzip)",
		"session_id": "00000000-0000-0000-0000-000000000000",
		"region": "Unimplemented",
		"latlon": "Unimplemented",
		"friend_count": ###,
		"country": "Unimplemented"
	},
	"events": [{
		"event_name": "APP_OPEN",
		"event_timestamp": 1423072534267,
		"event_params": {
			"open_state": "NORMAL"
		}
	}],
	"batch_id": "CENSORED"
}

**Comment by mgp25** [comment]
Here are some captures (login, logout, analitic events...) iOS: http://cl.ly/2R2F0K0h3p0W

I recommend for the android users go to `/data/data/com.snapchat.android` and check for the `sharedPrefs` and `gaClientId`

```xml
 ga<?xml version='1.0' encoding='utf-8' standalone='yes' ?>
<map>
<string name="birthday">myBirthday</string>
<string name="phone_number">+myPhone</string>
<int name="num_snaps_sent" value="10" />
<boolean name="informationSavedInDatabase" value="true" />
<string name="lastSuccessfulLoginUsername">myUser</string>
<string name="currentUser">myUser</string>
<null name="PHOTO_EFFECT_TASK_BREADCRUMB" />
<int name="score" value="15" />
<int name="story_privacy_setting" value="1" />
<boolean name="never_show_post_story_dialog" value="false" />
<boolean name="is_logged_in" value="true" />
<int name="version" value="210" />
<string name="username">myUser</string>
<null name="gcm_registration_id" />
<string name="mobile_verification_key">NjcxODp**************W5n</string>
<int name="num_snaps_received" value="5" />
<int name="num_best_friends" value="3" />
<boolean name="smart_filters_setting" value="false" />
<boolean name="front_facing_flash_setting" value="false" />
<boolean name="special_text_setting" value="false" />
<boolean name="snapchatUserIsLoggedIn" value="true" />
<boolean name="visual_filters_setting" value="false" />
<boolean name="FLASH_ENABLED" value="false" />
<string name="auth_token">ca804b2d*************773a9158590</string>
<string name="sendersSinceLastOpened"></string>
<boolean name="is_searchable_by_phone_number" value="true" />
<int name="snap_privacy_setting" value="1" />
<boolean name="can_view_mature_content" value="false" />
<string name="email">******@**********.com</string>
<boolean name="is_snapkidz_account" value="false" />
<int name="snapsSinceLastOpened" value="0" />
<long name="last_seen_added_me_timestamp" value="1422624600707" />
<boolean name="replay_setting" value="false" />
<string name="user">{}</string>
</map>
```

**Comment by Snowlav** [comment]
any progress on the ios login? is there anything you guys might need in terms of tools etc?

**Comment by LinoSuarez** [comment]
About to give up, I just can't find how "X-Snapchat-Client-Token" is generated.

**Comment by Snowlav** [comment]
Don't give up! :o
I was thinking why not use the old api to login and then export the auth token to use with the new api.
When updating snapchat I never get asked to re-login, so that idea should work. Right ?

**Comment by LinoSuarez** [comment]
The old api still works, the problem is that it will get discontinued so that's why we're trying to update to the new version of snapchat before the old api get obsolete (:

**Comment by LinoSuarez** [comment]
Ghhhh I will just brute force the dam Snapchat Client Token, any ideas of what kind of encryption its used on that variable? 2 Titan Z + GPU Bruteforce software = can crack anything in hours

**Comment by Snowlav** [comment]
I have no clue but I like your plan, keep us updated. Nice graphic cards
dude :D

**Comment by KyleBoyer** [comment]
> Ghhhh I will just brute force the dam Snapchat Client Token, any ideas of what kind of encryption its used on that variable? 2 Titan Z + GPU Bruteforce software = can crack anything in hours

Mine looks like it is "v1:" then a hash. This hash looks like it could be one of the following methods:
- md5($salt.$pass.$salt)
- md5($salt.md5($pass))
- md5($salt.md5($pass.$salt))
- md5($salt.md5($salt.$pass))
- md5($username.0.$pass)
- md5(md5($pass).md5($salt))
- md5(md5($salt).$pass)
- HMAC-MD5 (key = $pass)
- HMAC-MD5 (key = $salt)
- md5($pass.md5($salt))
- WebEdition CMS
- IPB2+, MyBB1.2+
- md5(unicode($pass).$salt)
- Domain Cached Credentials2, mscash2
- md5($salt.unicode($pass))
- vBulletin > v3.8.5
- DCC2
- md5(md5($pass).$salt), vBulletin < v3.8.5

**Comment by liamcottle** [comment]
Format is v1:HASH:ANOTHER_HASH

**Comment by KyleBoyer** [comment]
Or rather, the colons probably separate different hashes.

So it looks like "v1:" and maybe a hash of the following types:
 - MD5
- NTLM
- MD4
- LM
- RAdmin v2.x
- Haval-128
- MD2
- RipeMD-128
- Tiger-128
- Snefru-128
- MD5(HMAC)
- MD4(HMAC)
- Haval-128(HMAC)
- RipeMD-128(HMAC)
- Tiger-128(HMAC)
- Snefru-128(HMAC)
- MD2(HMAC)
- MD5(ZipMonster)
- MD5(HMAC(Wordpress))
- Skein-256(128)
- Skein-512(128)
- md5($pass.$salt)
- md5($pass.$salt.$pass)
- md5($pass.md5($pass))
- md5($salt.$pass)
- md5($salt.$pass.$salt)
- md5($salt.$pass.$username)
- md5($salt.'-'.md5($pass))
- md5($salt.md5($pass))
- md5($salt.md5($pass).$salt)
- md5($salt.MD5($pass).$username)
- md5($salt.md5($pass.$salt))
- md5($salt.md5($salt.$pass))
- md5($salt.md5(md5($pass).$salt))
- md5($username.0.$pass)
- md5($username.LF.$pass)
- md5($username.md5($pass).$salt)
- md5(1.$pass.$salt)
- md5(3 x strtoupper(md5($pass)))
- md5(md5($pass)), Double MD5
- md5(md5($pass).$pass)
- md5(md5($pass).$salt), vBulletin < v3.8.5
- md4($salt.$pass)
- md4($pass.$salt)md5(md5($pass).md5($pass))
- md5(md5($pass).md5($salt))
- md5(md5($salt).$pass)
- md5(md5($salt).md5($pass))
- md5(md5($username.$pass).$salt)
- md5(md5(base64_encode($pass)))
- md5(md5(md5($pass)))
- md5(md5(md5(md5($pass))))
- md5(md5(md5(md5(md5($pass)))))
- md5(sha1($pass))
- md5(sha1(base64_encode($pass)))
- md5(sha1(md5($pass)))
- md5(sha1(md5($pass)).sha1($pass))
- md5(sha1(md5(sha1($pass))))
- md5(strrev($pass))
- md5(strrev(md5($pass)))
- md5(strtoupper(md5($pass)))
- md5(strtoupper(md5(strtoupper(md5(strtoupper(md5($pass)))))))
- strrev(md5($pass))
- strrev(md5(strrev(md5($pass))))
- 6 x md5($pass)
- 7 x md5($pass)
- 8 x md5($pass)
- 9 x md5($pass)
- 10 x md5($pass)
- 11 x md5($pass)
- 12 x md5($pass)	
and ":" and then another of maybe the following types: 
- SHA-384
- Keccak-384
- SHA-384(HMAC)
- sha384($salt.$pass)
- sha384($pass.$salt)
- Skein-512(384)
- Skein-1024(384)

**Comment by cuonic** [comment]
Hash 1 is 32 characters long, it's clearly an MD5. We need some more requests from iOS apps to look for patterns. Create a fake account, logout and login whilst saving the request and post everything here (headers / request) for everything up to login

**Comment by LinoSuarez** [comment]
Impossible to brute force, it's a really long String.

**Comment by liamcottle** [comment]
No ones had any luck with the iOS reversing?

**Comment by KyleBoyer** [comment]
> No ones had any luck with the iOS reversing?

Using decrypted Snapchat.app that I downloaded, I'm only able to see compiler code in Hopper which is extremely difficult for me. Does anyone have more experience with compiler code?

**Comment by WaterSpice** [comment]
I'm getting issues with using JorgenPhi's API. Everything works fine, but after a few hours I get the:
"Stop using 3rd party apps"

Anyway to fix this? Or anyway to spoof it?

**Comment by KyleBoyer** [comment]
> I'm getting issues with using JorgenPhi's API. Everything works fine, but after a few hours I get the:
> "Stop using 3rd party apps"
> 
> Anyway to fix this? Or anyway to spoof it?

The files on the repo are old. Try using these instead:
https://gist.github.com/cuonic/449a34af812fefc72522
https://gist.github.com/cuonic/bb106e1ce978ebf71c6e

Without the cache script, but with the crt file.

**Comment by WaterSpice** [comment]
@KyleBoyer 
Tried with new files.
This is my code:
http://pastebin.com/Vu3rFCwJ

Getting the error:
Trying to get property of non-object 
	Snapchat->getFriends( )	..\test.php:5

When I use those 2 files you linked me, it doesn't work.

When I put back JorgenPhi's repo's files, it works fine.

Any ideas?

**Comment by Snowlav** [comment]
@WaterSpice  could you make a separate issue for this?

**Comment by LinoSuarez** [comment]
It still works for me! If you want I can help you, skype me: lino.suarez.diaz
![screenshot 2015-02-05 19 25 11](https://cloud.githubusercontent.com/assets/6445704/6074144/3382c772-ad6d-11e4-870b-12abf0f061d0.png)

**Comment by KyleBoyer** [comment]
On the latest iOS Snapchat update(Snapchat/9.1.1.0 (iPhone6,1; iOS 8.1.3; gzip)), the X-Snapchat-Client-Token is only passed when logging in.

Hopefully that will simplify attempting to figure out what the hashes are. Also, once the login request is made, and an auth_token is returned, I was able to use the auth_token simultaneously on the authentic app, and with this PHP script(with the 8.1.1 useragent).

**Comment by LinoSuarez** [comment]
Guys I have an idea!!! What if we use Ajax to manage all the requests. Meaning php code doing all encryptions/decryptions then passing the data to the client and then the client will do the request instead of the server doing it. This way we could solve many banning problems, will be less suspicious since we will be using the client IP to make the requests

**Comment by KyleBoyer** [comment]
That could be a possibility, however some people use this PHP script to make a snapchat bot that runs via a cron job on a server, so the AJAX would make no difference in regards to IP...

**Comment by Maraumax** [comment]
@LinoSuarez We can not because of [CORS](http://en.wikipedia.org/wiki/Cross-origin_resource_sharing).

**Comment by JorgenPhi** [comment]
Anyway to bypass that? I mean, it's up to the browser to respect that.

**Comment by cuonic** [comment]
The problem is not the IP, well it is if you are blocked from using the 
Snapchat API, but that's another story.
Loads of people (different accounts) connect from the same IP at the 
same time at loads of places (Public WIFI hotspots, University campus, 
libraries...) so Snapchat do not detect third party apps because you 
login to multiple accounts with only one IP.

Using javascript or any other scripting language will not solve this 
problem, the problem is the requests you send to the API, the real 
application has a certain network pattern when it logs on :

1. It logs in with /loq/login using a valid ptoken, a correct dtoken1 
and dsig and providing the Authorization header.
2. The application fetches non viewed snaps via /ph/blob
3. The application fetches some story thumbnails
4. The application requests /loq/all_updates
5. The application posts data to sc-analytics.appspot.com/post_events
6. The application fetches some story thumbnails
7. The application posts the GCM device token to /ph/device
8. The application gets all conversations with a request to 
/loq/conversations
9. The application gets the user's snaptag with /bq/snaptag_download
10. The application starts fetching story blobs with requests to 
/bq/story_blob
11. The application fetches the "discover" channel list with a request 
to /discover/channel_list
12. The application fetchs all "discover" related trash
13. The application posts data to sc-analytics.appspot.com/analytics_events

That is one whole login process for the standard android application. So 
when you use the API with a third party client and do just /loq/login 
and instantly start sending snaps, alarm bells go off at Snapchat HQ, 
your network pattern doesn't match the usual one, you get locked out.

There are also other easy giveaways for third party applications : The 
User-Agent, the headers, the non usage of Gzip, the fact that by default 
snapchat_agent.php sends the "version" in the post data is useless 
(https://github.com/JorgenPhi/php-snapchat/blob/master/src/snapchat_agent.php#L289) 
and is also probably a tell tale sign that you're not using the correct 
application.

**Comment by Snowlav** [comment]
@cuonic what you are saying makes perfect sense, however what doesn't is that snapchat won't block the network when you are just using /loq/login and send snaps (for example) if just a few accounts are active, but it starts blocking when there's more accounts active on the same ip all doing the same process.

**Comment by cuonic** [comment]
@snowlav No I don't think that's the problem, I've only ever used one 
account and it was blocked because of logging in and instantly sending 
snaps. The detection is not instantaneous, it takes time, no idea how much.

**Comment by Snowlav** [comment]
@cuonic  that's really strange the only time I had an IP blocked on <5 accounts was when my code bugged out and it kept logging into one account over and over again (calling /loq/login a lot of times in a row) or when sending snaps one by one without a delay.

**Comment by WaterSpice** [comment]
@Snowlav @cuonic 
I changed the version and line 52
	public static $CURL_OPTIONS = array(

I changed those 2 and haven't gotten anymore warnings.

Also pretty off topic, but do you know if there's a way to see all users who viewed your story with the API? I can't figure out how to get your own story's data

**Comment by cuonic** [comment]
@snowlav Maybe the third party app is detected faster on multiple 
accounts because the same non standard network patterns are coming from 
the same IP, maybe there is a counter, and with multiple accounts you 
max out the counter faster.

**Comment by Snowlav** [comment]
I think a network isn't supposed to get banned for the actual 3rd party usage but rather what you do with it.
For example I never heard of public wifi spots getting banned due to 6snap users or any other windows phone 3rd party applications.

I think what blocks networks (may it be one or more accounts) is that activity performed (exactly as the message says when your network is blocked -suspicious activity-)
Which are; logging in A LOT on the same account, sending snaps with a nowhere human like speed etc.
Maybe every account on the same network having the exact same headers, useragent etc?

This is no joke but I even got my home IP banned with the use of NO 3rd party applications by switching from my android to ipod a couple of times on the same account.
(though this was a long time ago, I'd say at least 6 months back)

No 3rd party code was ever used on my network before that.

**Comment by cuonic** [comment]
Network blocking is something recent though, 6snap and other third party 
apps for phones published on some market (Google play, App store) never 
resulted in a network ban, and neither did anything. Only recently have 
Snapchat started to crack down on third party apps and performing 
network blocking.

**Comment by Snowlav** [comment]
I got network blocked before while just swapping from ipod to android and other way around as stated in my post.
It's not something they just started doing, the frequency might have become higher but it's certainly not just starting.

**Comment by LinoSuarez** [comment]
Funny SaveSnap it still using "Snapchat/4.1.07 (Nexus 4; Android 18; gzip)"

**Comment by mgp25** [comment]
Managed to trick snapchat for iOS with a tweak and make it work with the API :) i'll try to do a better tool with UI and upload to a cydia repo. Also i have the snapchat api using android data working, will send more information when possible.

Have a nice week.

Regards

**Comment by liamcottle** [comment]
Just thought I'd let you all know, Snapchat is getting their lawyers to send out cease and desist notices...
So this could be the end...

**Comment by LinoSuarez** [comment]
But to who?

**Comment by LinoSuarez** [comment]
Also It's possible to make this legal, what about all the snapchat hack apps on the app store?

**Comment by liamcottle** [comment]
First one was to the guy who built SnapMod:
http://forum.xda-developers.com/showpost.php?p=58764245&postcount=395

I believe they're in the process, (continuing to) remove all third party snapchat apps from all app stores.

I guess we could just remove the keys etc, and it would be legal?

**Comment by LinoSuarez** [comment]
lawyers, theyre the answer

**Comment by JorgenPhi** [comment]
I've yet to get anything. And if anything we can move this project over to
i2p or something else as long as there interest.

**Comment by JorgenPhi** [comment]
*there is. This reminds me. We need an IRC channel. What's your preferred
network?

**Comment by liamcottle** [comment]
I haven't used IRC, so anything is fine for me I guess?

**Comment by WaterSpice** [comment]
@JorgenPhi 
Just get a channel on Esper?

**Comment by JorgenPhi** [comment]
Sure. Let me add that to my bouncer.

Channel is #php-snapchat on irc.esper.net

**Comment by LinoSuarez** [comment]
I think snapchat is reading all this and yeah It can take down the public API. What about making a Skype group? I can also host the GIT on my private Gitlab server if needed :)

**Comment by JorgenPhi** [comment]
That'd be lovely. If someone wants to make a Skype group, let me know. The
irc channel is up and running, so if you want to talk in there, feel free.

Also, about the git, if it comes down to it, that would be great!

**Comment by WaterSpice** [comment]
@JorgenPhi Can you make the IRC public? Or invite "WaterSpice"
Me an @LinoSuarez can get the skype group rolling if you want, I have him on skype.

**Comment by JorgenPhi** [comment]
It's not public? And yes I'll invite you.

**Comment by JorgenPhi** [comment]
My bad. Accidentally had it set +i.

**Comment by LinoSuarez** [comment]
@JorgenPhi  add me to skype: lino.suarez.diaz
I already added 3 people to the group

**Comment by JorgenPhi** [comment]
Alright. Follow up. The irc channel exists, but is fairly dead. Skype group
is much more popular. If you want an invite, let me know or add Lino and
he'll add you to the group.

**Comment by ashokfernandez** [comment]
@LinoSuarez - I've sent a request to you on Skype!

**Comment by liamcottle** [comment]
lol! They've changed the user agent again...

New format: `Snapchat/9.1.0.0 (SM-G900I; Android 4.4.2#G900IDVU1ANK1#19; gzip)`

... and, even more interesting, is this user agent returns ALL snap blobs unencrypted... I guess they realised the whole encrypting and decrypting snaps was pointless if we all know the encryption/decryption keys...

**Comment by JorgenPhi** [comment]
All blobs are unencrypted?

*parties internally*

**Comment by liamcottle** [comment]
I don't think story blobs are unencrypted, though I highly doubt they'd unencrypt those...

However, I just checked the old user agent, and that returns unencrypted blobs as well...

So they must be working on the API right now or something, because they were all encrypted about 20 minutes ago, and now they're not, so the server must pre-decrypt them all now..

... and I guess the new update to the snapchat app will stop encrypting them.

**Comment by liamcottle** [comment]
I was working on some things, and all of a sudden decryption was failing... so looked at blob data, not even encrypted...

**Comment by LinoSuarez** [comment]
Some time ago they says they where going to release an API maybe something
of that is going on

**Comment by liamcottle** [comment]
Could be... But I highly doubt they're going to even release an API...

Just all talk to keep us waiting... Because the whole point of snapchat is to send media that will be gone shortly after, and if they released a public API to download snaps, then it would make their whole system pointless.

**Comment by liamcottle** [comment]
Looking into it a bit more, I reckon snapchat is sending ping requests to their server via their chat socket... and flagging accounts that's don't ping... Could be something to look into?

**Comment by KyleBoyer** [comment]
Also when sending data, through charles proxy it looks like there is an extra parameter under data as seen here: http://s10.postimg.org/66bhe8ovd/namedata.png
Notice: filename="data"

**Comment by realmuster** [comment]
@LinoSuarez can you please accept my friend request on skype? ;)

**Comment by Spurlex** [comment]
@JorgenPhi Can you add Xelrups to the group thx

**Comment by mib4fun** [comment]
@JorgenPhi  add me too. **bellamy-fan**

**Comment by iguanaman** [comment]
@LinoSuarez sent you a skype request with details, thanks

**Comment by andresarocho** [comment]
@LinoSuarez Can you please add volcom526

**Comment by adriancooney** [comment]
@LinoSuarez add me too? cooney.adrian. Cheers for the good work guys!

**Comment by KiranPanesar** [comment]
@LinoSuarez Can you add me to the group? kiranspanesar. This thread is really interesting.

**Comment by jalandro** [comment]
See Gitter for chat group at github. https://gitter.im/

@JorgenPhi
Please consider making a team with @cuonic and @mgp25 (They have great forks).

**Comment by niober** [comment]
Okey I think you will be really interested in what I have to say here. I have ALMOST solved the whole snapchat downgrade problem. I can tell you all about how but first in need someone's help here. Now I have successfully installed snapchat 6.1.1 on my jailbroken iPhone 4 ios 7.1.2. When I open snapchat 6.1.1 there is no crash and I can successfully watch my friends story's and also load story's. The problem is that I can't upload story's, I can't load snaps that I got and I can't send snaps either. So If someone know how to fix that problem then I think we have solved the whole problem;)

**Comment by KiranPanesar** [comment]
@niober This is using the Android API?

**Comment by vivijimmy** [comment]
@cuonic Hey, in your code 'Authorization: AidLogin 3976909495270287464:285036945524349568', I know the 3976909495270287464 should be the Android_id, but what is that 285036945524349568? Thanks!

**Comment by cuonic** [comment]
I obtained it by sniffing the GCM registration request, but according to [this document](https://www.cs.indiana.edu/~zhou/files/mobile_cloud.pdf) (Page 4) the format is `android-id:device-token`.

**Comment by vivijimmy** [comment]
@cuonic Hmm, I used uploadDeviceConfig and got the device token something like 1440513696098*blahblah, which seems not the number after the android_id. I read the same document too and I knew it said that number was device token. Were you using uploadDeviceConfig? Or "checkin" API?

**Comment by cuonic** [comment]
That's more a timestamp than a device token. I didn't use either of them endpoints, where is uploadDeviceConfig from ?

**Comment by vivijimmy** [comment]
@cuonic I tried to sniff all the data out and in for Google Play application and I found it could only send this Authorization: AidLogin android_id:unknown_number. In all the data comes back from Google Play I can not find out the same number as the unknown_number after android_id here. I think it should be something Google Play application generated but really curious about what that is... Tried to search all over the Internet but it is very little mentioned.

**Comment by cuonic** [comment]
I'm looking into it now, the second item is actually called `CheckinTask_securityToken`

**Comment by vivijimmy** [comment]
@cuonic Great! Many thanks bro! Wow. This CheckinTask_securityToken seems to be written into the CheckinService.xml by Google Play Service. I will find out which writes this and keep you updated.

**Comment by cuonic** [comment]
the `android_id` and `security_token` are provided to the device on their first `checkin` request (`https://android.clients.google.com/checkin`). Here is an example of what is sent to the endpoint: https://github.com/nviennot/android-checkin. I'm creating a tool that performs all the required requests in Python at the moment, but this is unrelated to Snapchat so contact me via email.

**Comment by devhex** [comment]
Has anyone figured out how the iOS authentication works yet or is it being discussed anywhere?

**Comment by mgp25** [comment]
@devhex use this: https://github.com/mgp25/SC-API

Wiki available with all documentation

Regards

**Comment by devhex** [comment]
Yeah I've seen it and it's great, but I want to find out how the X-Snapchat-Client-Auth-Token is generated on iOS without the need of a Google account.

**Comment by mgp25** [comment]
@devhex you will need to reverse iOS app, use IDA or hopper

**Comment by nguyenngoccuong** [comment]
Is there any solution to generate X-Snapchat-Client-Auth-Token and X-Snapchat-Client-Token, please help me, thanks, my skype: siverrally

**Comment by filfat** [comment]
Bump the above comment

**Comment by mzaboss** [comment]
Full Snapchat API using a new and complex tweak.

You can use it to:

Create snapchat bots.

* Features:

# Send a media to a story
# Send a snap
# Add a friend
# Delete a friend
# Mark a snap video as viewed

Create a third app on the App Store or Play Store.

more dev informations will be sent in PDF.

Monthly subscription: $450! only. Unlimited requests to snapchat.

FREE 24h trial!

Skype: mzaboss

Thank you,
regards.
